runtime/sam/expr/function/parse.go

```diff
@@ -1,79 +1,119 @@
+	zctx    *super.Context
+	typ     super.Type
+	builder zcode.Builder
+}
+
+func newParseURI(zctx *super.Context) *ParseURI {
+	return &ParseURI{zctx: zctx, typ: URIType(zctx)}
 }
 
 func (p *ParseURI) Call(_ super.Allocator, args []super.Value) super.Value {
-	in := args[0]
-	if !in.IsString() || in.IsNull() {
-		return p.zctx.WrapError("parse_uri: non-empty string arg required", in)
+	in := args[0].Under()
+	if !in.IsString() {
+		return p.zctx.WrapError("parse_uri: string arg required", args[0])
+	}
+	if in.IsNull() {
+		return super.NewValue(p.typ, nil)
+	}
+	p.builder.Reset()
+	err := EncodeURI(&p.builder, in.AsString())
+	if err != nil {
+		return p.zctx.WrapError(err.Error(), args[0])
 	}
-	s := super.DecodeString(in.Bytes())
+	return super.NewValue(p.typ, p.builder.Bytes())
+}
+
+func URIType(zctx *super.Context) super.Type {
+	const uriType = `{scheme:string,opaque:string,user:string,password:string,host:string,port:uint16,path:string,query:|{string:[string]}|,fragment:string}`
+	typ, err := zson.ParseType(zctx, uriType)
+	if err != nil {
+		panic(err)
+	}
+	return typ
+}
+
+func EncodeURI(b *zcode.Builder, s string) error {
 	u, err := url.Parse(s)
 	if err != nil {
-		return p.zctx.WrapError("parse_uri: "+err.Error(), in)
-	}
-	var v struct {
-		Scheme   *string    `zed:"scheme"`
-		Opaque   *string    `zed:"opaque"`
-		User     *string    `zed:"user"`
-		Password *string    `zed:"password"`
-		Host     *string    `zed:"host"`
-		Port     *uint16    `zed:"port"`
-		Path     *string    `zed:"path"`
-		Query    url.Values `zed:"query"`
-		Fragment *string    `zed:"fragment"`
+		return fmt.Errorf("parse_uri: %w", err)
 	}
 	if u.Scheme != "" {
-		v.Scheme = &u.Scheme
+		b.Append([]byte(u.Scheme))
+	} else {
+		b.Append(nil)
 	}
 	if u.Opaque != "" {
-		v.Opaque = &u.Opaque
+		b.Append([]byte(u.Opaque))
+	} else {
+		b.Append(nil)
 	}
 	if s := u.User.Username(); s != "" {
-		v.User = &s
+		b.Append([]byte(s))
+	} else {
+		b.Append(nil)
 	}
 	if s, ok := u.User.Password(); ok {
-		v.Password = &s
+		b.Append([]byte(s))
+	} else {
+		b.Append(nil)
 	}
 	if s := u.Hostname(); s != "" {
-		v.Host = &s
+		b.Append([]byte(s))
+	} else {
+		b.Append(nil)
 	}
 	if portString := u.Port(); portString != "" {
 		u64, err := strconv.ParseUint(portString, 10, 16)
 		if err != nil {
-			return p.zctx.WrapError("parse_uri: invalid port: "+portString, in)
+			return fmt.Errorf("parse_uri: invalid port %q", portString)
 		}
-		u16 := uint16(u64)
-		v.Port = &u16
+		b.Append(super.EncodeUint(u64))
+	} else {
+		b.Append(nil)
 	}
 	if u.Path != "" {
-		v.Path = &u.Path
+		b.Append([]byte(u.Path))
+	} else {
+		b.Append(nil)
 	}
 	if q := u.Query(); len(q) > 0 {
-		v.Query = q
+		var mb zcode.Builder
+		mb.BeginContainer()
+		for k, list := range q {
+			mb.Append([]byte(k))
+			mb.BeginContainer()
+			for _, v := range list {
+				mb.Append([]byte(v))
+			}
+			mb.EndContainer()
+		}
+		mb.EndContainer()
+		b.Append(super.NormalizeMap(mb.Bytes().Body()))
```

Comment:
Instead of creating another zoce.Builder like this, use zcode.Builder.TransformContainer. 
```suggestion
		b.BeginContainer()
		for k, list := range q {
			b.Append([]byte(k))
			b.BeginContainer()
			for _, v := range list {
				b.Append([]byte(v))
			}
			b.EndContainer()
		}
		b.TransformContainer(super.NormalizeMap)
		b.EndContainer()
```